tests/linen/module_test.py

```diff
@@ -590,6 +590,40 @@ def __call__(self, x):
     trace = mlp.apply(variables, x)
     self.assertEqual(trace, expected_trace)
 
+  def test_call_unbound_compact_module_methods(self):
+    dense = Dense(3)
+    with self.assertRaisesRegex(ValueError, "compact.*unbound module"):
+      dense(jnp.ones((1, )))
+
+
+  def test_call_unbound_noncompact_module_methods(self):
+    class EmptyModule(nn.Module):
+      foo: int = 3
+
+      def bar(self):
+        return self.foo
+
+    empty = EmptyModule()
+    # It's fine to call methods of unbound methods that don't depend on
+    # attributes defined during `setup`
+    self.assertEqual(empty.bar(), 3)
+
+
+  def test_call_unbound_noncompact_module_method_without_setup(self):
+    class EmptyModule(nn.Module):
+      def setup(self):
+        self.setup_called = True
+
+      def bar(self):
+        return self.setup_called
```

Comment:
I think what we really want here is a test that does something like `self.param(...)` and throws a proper error

Reply:
Yeah good point, we should add another test for that and all the other methods.

Reply:
But the error you'd get there should say "can't define submodules or parameters in a non-compact method". But agreed we should have tests.

Reply:
Ah yes for variables you're right. For `make_rng` and `variables()` you still get the None issue because they don't have tot be in compact or setup

Reply:
Ah you're right, I'll add tests for all these cases and make sure the errors are clear.

Reply:
OK done -- pushed a new commit that fixes the three remaining cases where `if scope is None` wasn't checked.